fill_in_the_blanks/closures.adoc

```diff
@@ -0,0 +1,104 @@
+= Fill In The Blanks: Closures
+:source-language: rust
+
+== Task
+Take the code snippet at the end of this sheet and complete it where indicated with `✅ TODO`.
+
+When you're done, it should print the following output:
+
+----
+found: hay
+found: hay
+found: hay
+found: needle
+found: hay
+found: hay
+top of the haystack: hay
+look, I found the needle: ["needle"]
+bale size: 5
+new haystack: [(), (), (), (), ()]
+----
+
+== Code snippet to be completed
+
+[source,rust]
+----
+struct BaleOfStraw {
+    size: usize
+}
+
+// NOTE: Once you're done, your ouput could look like this:
+//
+// found: hay
+// found: hay
+// found: hay
+// found: needle
+// found: hay
+// found: hay
+// top of the haystack: hay
+// look, I found the needle: ["needle"]
+// bale size: 5
+// new haystack: [(), (), (), (), ()]
+
+fn main(){
+    // RECAP: ANATOMY OF A CLOSURE
+    //============================
+    //
+    let rummage = | element | {         // 👈  input parameters MAY omit type annotation
+                                        //     if type can be inferred. The input parameter
+                                        //     is also referred to as the "captured value".
+        println!("found: {}", element);
+                                        // 👈  no return statement: closures *can* return values,
+                                        //     but this one doesn't
+    };                                  // 👈  {}s are only needed for multi-line closures
+
+    let haystack = vec!["hay", "hay", "hay", "needle", "hay", "hay"];
+
+    // REFERENCING CLOSURES
+    //=====================
+    // *By default*, closures capture variables by reference. When given a choice, this is the
+    // behaviour inferred by the compiler.
+
+    haystack.iter().for_each(rummage);
+    // note that we can directly borrow the contents of `haystack` after using the `rummage`
+    // closure, because it only borrowed the elements (and is now done)
+    println!("top of the haystack: {}", haystack[0]);
+
+    // MUTATING CLOSURES
+    //==================
+    // Closures can mutate the variables they are capturing
+
+    // 👀  Closures can be used as function arguments.
+
+    // ✅ TODO: remove all the hay from `haystack` by checking whether `key` is a needle
+    let mut haystack_clone = haystack.clone();
+    haystack_clone.retain(|key| /* check key here */ );
+    println!("look, I found the needle: {:?}", haystack_clone);
+
+    // 👀  a common use case for closures is to transform collections
+    //     using e.g. `map()` and `filter()`.
+
+    // ✅ TODO: use `map()` to increase the bale size with every "hay" in the haystack
+    let mut bale = BaleOfStraw{ size: 0 };
+
+    let new_haystack: Vec<_> = haystack
+        .into_iter()
+        .filter(|element | *element == "hay")
+        .map( /* increment bale size here */ )
```

Comment:
if the *only* intent is to increment bale size, change `map` to `inspect`

Reply:
the intent is to introduce people to map :D how about we change this to...
```suggestion
        .map( /* remove straws and increment bale size here */ )
```
?

Reply:
what code do you have in mind for "remove straws"?

Reply:
I was looking for something like:

```
    let new_haystack: Vec<_> = haystack
        .into_iter()
        .filter(|element| *element == "hay")
        .map(|element| bale.size += 1)
        .collect();
```

I'll also add an example solution